internal/controller/numaflowcontroller/numaflowcontroller_controller.go

```diff
@@ -258,33 +259,52 @@ func (r *NumaflowControllerReconciler) reconcile(
 	}
 
 	newVersion := controller.Spec.Version
-	newVersionTargetObjs, err := determineTargetObjects(controller, newVersion)
+	newVersionTargetObjsMap, newVersionTargetObjs, err := determineTargetObjects(controller, newVersion)
 	if err != nil {
 		return ctrl.Result{}, fmt.Errorf("unable to determine the target objects for the new version %s: %w", newVersion, err)
 	}
 
-	numaLogger.Debugf("found %d target objects associated with NumaflowController version %s", len(newVersionTargetObjs), newVersion)
+	numaLogger.Debugf("found %d target objects associated with the new NumaflowController version %s", len(newVersionTargetObjs), newVersion)
 
-	// If a Deployment already exists for the NumaflowController and the new version differs from the current Deployment version,
-	// delete all the NumaflowController child resources and recreate them at the next reconciliation
 	if deploymentExists {
 		currentVersion, err := getControllerDeploymentVersion(existingDeployment)
 		if err != nil {
 			return ctrl.Result{}, fmt.Errorf("error while getting the version of the existing NumaflowController Deployment: %w", err)
 		}
 
-		childResourcesNeedToBeDeleted := controller.Spec.Version != currentVersion
-		if childResourcesNeedToBeDeleted {
-			numaLogger.Debugf("current Deployment image tag=%q differs from desired %q", currentVersion, controller.Spec.Version)
+		currentTargetObjsMap, _, err := determineTargetObjects(controller, currentVersion)
```

Comment:
(for this reason, I think just executing the "else" block would be fine, since ultimately we want to make sure that the resources we're attempting to create are deleted first)